### PR TITLE
Fix placement of annotation items in WMS GetMap

### PR DESCRIPTION
**Pull request by mhugent**
At the moment, annotation items (e.g. text annotation, svg annotation, ...) are always drawn at position (0,0). If an item is in a QGraphicsScene, this is ok, because the item draws in it's local coordinate system. In WMS GetMap we have no graphics scene, so the painter needs to be translated to make the items appear at the correct location in the map.

### CODE REVIEW

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 1]
```diff
@@ -3265,7 +3265,32 @@ namespace QgsWms
...
+      //consider item position
+      double offsetX = 0;
+      double offsetY = 0;
+      if ( annotation->hasFixedMapPosition() )
+      {
+        QgsPointXY mapPos = annotation->mapPosition();
+        if ( mapSettings.destinationCrs() != annotation->mapPositionCrs() )
+        {
+          QgsCoordinateTransform coordTransform( annotation->mapPositionCrs(), mapSettings.destinationCrs(), mapSettings.transformContext() );
+          mapPos = coordTransform.transform( mapPos );
```
Comment (nyalldawson) [inline src/server/services/wms/qgswmsrenderer.cpp:3277]: better catch QgsCsException here, just to be safe

**Comment by nyalldawson** [comment]
Changes look good to me :+1:

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 2]
```diff
@@ -3265,7 +3265,32 @@ namespace QgsWms
...
+        }
+        QgsPointXY devicePos = mapSettings.mapToPixel().transform( mapPos );
```
Comment (pblottiere) [inline src/server/services/wms/qgswmsrenderer.cpp:3279]:
```suggestion
        const QgsPointXY devicePos = mapSettings.mapToPixel().transform( mapPos );
```

File: src/server/services/wms/qgswmsrenderer.cpp [hunk 3]
```diff
@@ -3265,7 +3265,32 @@ namespace QgsWms
...
+        offsetX = devicePos.x();
+        offsetY = devicePos.y();
+      }
+      else
+      {
+        QPointF relativePos = annotation->relativePosition();
```
Comment (pblottiere) [inline src/server/services/wms/qgswmsrenderer.cpp:3285]:
```suggestion
        const QPointF relativePos = annotation->relativePosition();
```

**Comment by pblottiere** [comment]
Thanks @mhugent, code looks good to me :+1:. Can you add some tests please?

**Review by nyalldawson — COMMENTED** [review]
Can you black out the white areas from the top left of the WMS_GetMap_Annotations_mask.png file? If these are left than the test would still pass even if your fix is broken by someone in future and we regress to the old top-left placement

**Comment by mhugent** [comment]
> Can you black out the white areas from the top left of the WMS_GetMap_Annotations_mask.png file? If these are left than the test would still pass even if your fix is broken by someone in future and we regress to the old top-left placement

Right, they are blacked out now. I have no idea how those white areas came into the mask image.

**Review by pblottiere — APPROVED** [review]

**Pull request merged**